## Rep 1 Python Script
#### Group Members
* Luis
* Diego
* Álvaro


In [80]:
pip install pyreadr==0.4.2

In [81]:
#Import packages
import pandas as pd
import numpy as np
import pyreadr as rr 
import math

In [82]:
#import .Rdata 
rdata_read = rr.read_r(r"../../data/wage2015_subsample_inference.Rdata")
#select data
data = rdata_read[ 'data' ]
print(data.shape)
print(type(data))

(5150, 20)
<class 'pandas.core.frame.DataFrame'>


In [83]:
#check data
data

wage     lwage  sex  shs  hsg  scl  clg   ad   mw   so   we  \
rownames                                                                     
10         9.615385  2.263364  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
12        48.076923  3.872802  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
15        11.057692  2.403126  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
18        13.942308  2.634928  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0   
19        28.846154  3.361977  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
...             ...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
32620     14.769231  2.692546  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0   
32624     23.076923  3.138833  1.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0   
32626     38.461538  3.649659  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0   
32631     32.967033  3.495508  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0   
32643     17.307692  2.851151  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0   

           ne  exp1  exp2    exp3     exp4   occ occ2   ind ind2  
rownames                                                          
10        1.0   7.0  0.49   0.343   0.2401  3600   11  8370   18  
12        1.0  31.0  9.61  29.791  92.3521  3050   10  5070    9  
15        1.0  18.0  3.24   5.832  10.4976  6260   19   770    4  
18        1.0  25.0  6.25  15.625  39.0625   420    1  6990   12  
19        1.0  22.0  4.84  10.648  23.4256  2015    6  9470   22  
...       ...   ...   ...     ...      ...   ...  ...   ...  ...  
32620     0.0   9.0  0.81   0.729   0.6561  4700   16  4970    9  
32624     0.0  12.0  1.44   1.728   2.0736  4110   13  8680   20  
32626     0.0  11.0  1.21   1.331   1.4641  1550    4  3680    6  
32631     0.0  10.0  1.00   1.000   1.0000  2920    9  6570   11  
32643     0.0  14.0  1.96   2.744   3.8416  1610    5  7460   14  

[5150 rows x 20 columns]

#### Variable description
* occ : occupational classification
* ind : industry classification
* lwage : log hourly wage
* sex : gender (1 female) (0 male)
* shs : some high school
* hsg : High school graduated
* scl : Some College
* clg: College Graduate
* ad: Advanced Degree
* ne: Northeast
* mw: Midwest
* so: South
* we: West
* exp1: experience

In [84]:
data.info()
#No null data, it's ok!

<class 'pandas.core.frame.DataFrame'>
Index: 5150 entries, 10 to 32643
Data columns (total 20 columns):
 #   Column  Non-Null Count  Dtype   
---  ------  --------------  -----   
 0   wage    5150 non-null   float64 
 1   lwage   5150 non-null   float64 
 2   sex     5150 non-null   float64 
 3   shs     5150 non-null   float64 
 4   hsg     5150 non-null   float64 
 5   scl     5150 non-null   float64 
 6   clg     5150 non-null   float64 
 7   ad      5150 non-null   float64 
 8   mw      5150 non-null   float64 
 9   so      5150 non-null   float64 
 10  we      5150 non-null   float64 
 11  ne      5150 non-null   float64 
 12  exp1    5150 non-null   float64 
 13  exp2    5150 non-null   float64 
 14  exp3    5150 non-null   float64 
 15  exp4    5150 non-null   float64 
 16  occ     5150 non-null   category
 17  occ2    5150 non-null   category
 18  ind     5150 non-null   category
 19  ind2    5150 non-null   category
dtypes: category(4), float64(16)
memory usage: 736.3+ KB


In [85]:
#Summary statistics
data.describe()

wage        lwage          sex          shs          hsg  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean     23.410410     2.970787     0.444466     0.023301     0.243883   
std      21.003016     0.570385     0.496955     0.150872     0.429465   
min       3.021978     1.105912     0.000000     0.000000     0.000000   
25%      13.461538     2.599837     0.000000     0.000000     0.000000   
50%      19.230769     2.956512     0.000000     0.000000     0.000000   
75%      27.777778     3.324236     1.000000     0.000000     0.000000   
max     528.845673     6.270697     1.000000     1.000000     1.000000   

               scl          clg           ad           mw           so  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean      0.278058     0.317670     0.137087     0.259612     0.296505   
std       0.448086     0.465616     0.343973     0.438464     0.456761   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       1.000000     1.000000     0.000000     1.000000     1.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

                we           ne         exp1         exp2         exp3  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean      0.216117     0.227767    13.760583     3.018925     8.235867   
std       0.411635     0.419432    10.609465     4.000904    14.488962   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     5.000000     0.250000     0.125000   
50%       0.000000     0.000000    10.000000     1.000000     1.000000   
75%       0.000000     0.000000    21.000000     4.410000     9.261000   
max       1.000000     1.000000    47.000000    22.090000   103.823000   

              exp4  
count  5150.000000  
mean     25.118038  
std      53.530225  
min       0.000000  
25%       0.062500  
50%       1.000000  
75%      19.448100  
max     487.968100

In [86]:
#Let's focus on some college and complete college educated individuals

Z = data [ ["lwage","sex","scl","clg","ad","ne","mw","so","we","exp1"] ]

data_scl = data[data[ 'scl' ] == 1 ]
Z_scl = data_scl[ ["lwage","sex","scl","clg","ad","ne","mw","so","we","exp1"] ]

data_clg = data[ data[ 'clg' ] == 1 ]
Z_clg = data_clg[ [ "lwage","sex","scl","clg","ad","ne","mw","so","we","exp1" ] ]

table = np.zeros( (10, 3) )
table[:, 0] = Z.mean().values
table[:, 1] = Z_scl.mean().values
table[:, 2] = Z_clg.mean().values
table_pandas = pd.DataFrame( table, columns = [ 'All', 'Some College Ed.', 'College Ed.'])
table_pandas.index = ["Log Wage","Sex","Some College","Collage Graduate","Advanced Degree", "Northeast","Midwest","South","West","Experience"]

table_pandas


All  Some College Ed.  College Ed.
Log Wage           2.970787          2.847335     3.133670
Sex                0.444466          0.453911     0.485941
Some College       0.278058          1.000000     0.000000
Collage Graduate   0.317670          0.000000     1.000000
Advanced Degree    0.137087          0.000000     0.000000
Northeast          0.227767          0.186453     0.261614
Midwest            0.259612          0.305866     0.231051
South              0.296505          0.282123     0.289120
West               0.216117          0.225559     0.218215
Experience        13.760583         14.682612    10.966381

In [87]:
# Difference in completed college v. some college workers

diff_mean= data_clg['lwage'].mean() - data_scl['lwage'].mean()

print("Comparing by means, being a completed college workers brings ", round(diff_mean*100, 2), "% more salary than a some college worker")

Comparing by means, being a completed college workers brings  28.63 % more salary than a some college worker


### Unconditional regression

In [88]:
#An unconditional comparision is generally biased. Let's make some regressions to apply some Econometrics to capture the causal effect of college education in wages.

#install pkgs
!pip install statsmodels

import statsmodels.api as sm
import statsmodels.formula.api as smf

In [89]:
#previously clean our dataset (some college v. college completed subset)
subset= data.copy()
subset

wage     lwage  sex  shs  hsg  scl  clg   ad   mw   so   we  \
rownames                                                                     
10         9.615385  2.263364  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
12        48.076923  3.872802  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
15        11.057692  2.403126  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
18        13.942308  2.634928  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0   
19        28.846154  3.361977  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
...             ...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
32620     14.769231  2.692546  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0   
32624     23.076923  3.138833  1.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0   
32626     38.461538  3.649659  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0   
32631     32.967033  3.495508  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0   
32643     17.307692  2.851151  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0   

           ne  exp1  exp2    exp3     exp4   occ occ2   ind ind2  
rownames                                                          
10        1.0   7.0  0.49   0.343   0.2401  3600   11  8370   18  
12        1.0  31.0  9.61  29.791  92.3521  3050   10  5070    9  
15        1.0  18.0  3.24   5.832  10.4976  6260   19   770    4  
18        1.0  25.0  6.25  15.625  39.0625   420    1  6990   12  
19        1.0  22.0  4.84  10.648  23.4256  2015    6  9470   22  
...       ...   ...   ...     ...      ...   ...  ...   ...  ...  
32620     0.0   9.0  0.81   0.729   0.6561  4700   16  4970    9  
32624     0.0  12.0  1.44   1.728   2.0736  4110   13  8680   20  
32626     0.0  11.0  1.21   1.331   1.4641  1550    4  3680    6  
32631     0.0  10.0  1.00   1.000   1.0000  2920    9  6570   11  
32643     0.0  14.0  1.96   2.744   3.8416  1610    5  7460   14  

[5150 rows x 20 columns]

In [90]:
subset['clg_comp'] = np.where(subset['clg']==1, 1, 0)

subset

wage     lwage  sex  shs  hsg  scl  clg   ad   mw   so  ...  \
rownames                                                               ...   
10         9.615385  2.263364  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...   
12        48.076923  3.872802  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...   
15        11.057692  2.403126  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...   
18        13.942308  2.634928  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  ...   
19        28.846154  3.361977  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...   
...             ...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
32620     14.769231  2.692546  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...   
32624     23.076923  3.138833  1.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  ...   
32626     38.461538  3.649659  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  ...   
32631     32.967033  3.495508  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...   
32643     17.307692  2.851151  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  ...   

           ne  exp1  exp2    exp3     exp4   occ occ2   ind ind2 clg_comp  
rownames                                                                   
10        1.0   7.0  0.49   0.343   0.2401  3600   11  8370   18        1  
12        1.0  31.0  9.61  29.791  92.3521  3050   10  5070    9        1  
15        1.0  18.0  3.24   5.832  10.4976  6260   19   770    4        0  
18        1.0  25.0  6.25  15.625  39.0625   420    1  6990   12        0  
19        1.0  22.0  4.84  10.648  23.4256  2015    6  9470   22        1  
...       ...   ...   ...     ...      ...   ...  ...   ...  ...      ...  
32620     0.0   9.0  0.81   0.729   0.6561  4700   16  4970    9        1  
32624     0.0  12.0  1.44   1.728   2.0736  4110   13  8680   20        0  
32626     0.0  11.0  1.21   1.331   1.4641  1550    4  3680    6        0  
32631     0.0  10.0  1.00   1.000   1.0000  2920    9  6570   11        0  
32643     0.0  14.0  1.96   2.744   3.8416  1610    5  7460   14        0  

[5150 rows x 21 columns]

In [92]:
#dispense with individual with no completed nor some college education
subset = subset[subset['shs']==0]
subset = subset[subset['hsg']==0]
subset = subset[subset['ad']==0]

#dispense with clg and scl columns
subset.drop(['shs','hsg','ad'], axis='columns', inplace=True)

subset

wage     lwage  sex  scl  clg   mw   so   we   ne  exp1  exp2  \
rownames                                                                       
10         9.615385  2.263364  1.0  0.0  1.0  0.0  0.0  0.0  1.0   7.0  0.49   
12        48.076923  3.872802  0.0  0.0  1.0  0.0  0.0  0.0  1.0  31.0  9.61   
19        28.846154  3.361977  1.0  0.0  1.0  0.0  0.0  0.0  1.0  22.0  4.84   
30        11.730769  2.462215  1.0  0.0  1.0  0.0  0.0  0.0  1.0   1.0  0.01   
71        19.230769  2.956512  1.0  0.0  1.0  0.0  0.0  0.0  1.0   4.0  0.16   
...             ...       ...  ...  ...  ...  ...  ...  ...  ...   ...   ...   
32596     45.546559  3.818735  1.0  0.0  1.0  0.0  0.0  1.0  0.0   5.0  0.25   
32606     24.038462  3.179655  1.0  0.0  1.0  0.0  0.0  1.0  0.0  17.0  2.89   
32619     13.846154  2.628007  0.0  0.0  1.0  0.0  0.0  1.0  0.0  10.0  1.00   
32620     14.769231  2.692546  0.0  0.0  1.0  0.0  0.0  1.0  0.0   9.0  0.81   
32624     23.076923  3.138833  1.0  1.0  0.0  0.0  0.0  1.0  0.0  12.0  1.44   

            exp3     exp4   occ occ2   ind ind2  clg_comp  
rownames                                                   
10         0.343   0.2401  3600   11  8370   18         1  
12        29.791  92.3521  3050   10  5070    9         1  
19        10.648  23.4256  2015    6  9470   22         1  
30         0.001   0.0001  1650    5  7460   14         1  
71         0.064   0.0256  3255   10  8190   18         1  
...          ...      ...   ...  ...   ...  ...       ...  
32596      0.125   0.0625  3255   10  8190   18         1  
32606      4.913   8.3521  2550    8  9480   22         1  
32619      1.000   1.0000   800    2   770    4         1  
32620      0.729   0.6561  4700   16  4970    9         1  
32624      1.728   2.0736  4110   13  8680   20         0  

[3068 rows x 18 columns]

#### Make regressions to find the causal effect of college education in wages.

In [94]:
#remember: ~ = Alt + 126

nocontrol_model = smf.ols(formula = 'lwage ~ clg_comp', data=subset )
nocontrol_est = nocontrol_model.fit().summary2().tables[1]['Coef.']['clg_comp']
nocontrol_est
nocontrol_se2 = nocontrol_model.fit().summary2().tables[1]['Std.Err.']['clg_comp']

# robust standar erros
HCV_coefs = nocontrol_model.fit().cov_HC0
nocontrol_se = np.power( HCV_coefs.diagonal() , 0.5)[1]
nocontrol_se

# print unconditional effect of gender and the corresponding standard error

print( f'The estimated college coefficient is {nocontrol_est} and the corresponding standard error is {nocontrol_se2}' )
print( f'The estimated college coefficient is {nocontrol_est} and the corresponding robust standard error is {nocontrol_se}','\n' )

The estimated college coefficient is 0.2863352099445873 and the corresponding standard error is 0.01901742129477679
The estimated college coefficient is 0.2863352099445873 and the corresponding robust standard error is 0.019018119623101924 



We obtain the unconditional college wage gap $28,63 \%$ is the same as the no control OLS regression estimator, $0.2863$